velox/exec/Window.cpp

```diff
@@ -0,0 +1,487 @@
+          windowNode->sources()[0]->outputType()->children(),
+          operatorCtx_->mappedMemory())),
+      decodedInputVectors_(numInputColumns_) {
+  auto inputType = windowNode->sources()[0]->outputType();
+  initKeyInfo(inputType, windowNode->partitionKeys(), {}, partitionKeyInfo_);
+  initKeyInfo(
+      inputType,
+      windowNode->sortingKeys(),
+      windowNode->sortingOrders(),
+      sortKeyInfo_);
+  allKeyInfo_.reserve(partitionKeyInfo_.size() + sortKeyInfo_.size());
+  allKeyInfo_.insert(
+      allKeyInfo_.cend(), partitionKeyInfo_.begin(), partitionKeyInfo_.end());
+  allKeyInfo_.insert(
+      allKeyInfo_.cend(), sortKeyInfo_.begin(), sortKeyInfo_.end());
+
+  std::vector<exec::RowColumn> inputColumns;
+  for (int i = 0; i < inputType->children().size(); i++) {
+    inputColumns.push_back(data_->columnAt(i));
+  }
+  // The WindowPartition is structured over all the input columns data.
+  // Individual functions access its input argument columns from it.
+  // The RowColumns are copied by the WindowPartition, so its fine to use
+  // a local variable here.
+  windowPartition_ = std::make_unique<WindowPartition>(
+      inputColumns, inputType->children(), pool());
+
+  createWindowFunctions(windowNode, inputType);
+}
+
+void Window::createWindowFunctions(
+    const std::shared_ptr<const core::WindowNode>& windowNode,
+    const RowTypePtr& inputType) {
+  auto fieldArgToChannel =
+      [&](const core::TypedExprPtr arg) -> std::optional<column_index_t> {
+    if (arg) {
+      std::optional<column_index_t> argChannel =
+          exprToChannel(arg.get(), inputType);
+      VELOX_CHECK(
+          argChannel.value() != kConstantChannel,
+          "Window doesn't allow constant arguments or frame end-points");
+      return argChannel;
+    }
+    return std::nullopt;
+  };
+
+  for (const auto& windowNodeFunction : windowNode->windowFunctions()) {
+    std::vector<TypePtr> argTypes;
+    std::vector<column_index_t> argIndices;
+    argTypes.reserve(windowNodeFunction.functionCall->inputs().size());
+    argIndices.reserve(windowNodeFunction.functionCall->inputs().size());
+    for (auto& arg : windowNodeFunction.functionCall->inputs()) {
+      argTypes.push_back(arg->type());
+      argIndices.push_back(fieldArgToChannel(arg).value());
+    }
+
+    windowFunctions_.push_back(WindowFunction::create(
+        windowNodeFunction.functionCall->name(),
+        argTypes,
+        argIndices,
+        windowNodeFunction.functionCall->type(),
+        operatorCtx_->pool()));
+
+    checkDefaultWindowFrame(windowNodeFunction);
+
+    windowFrames_.push_back(
+        {windowNodeFunction.frame.type,
+         windowNodeFunction.frame.startType,
+         windowNodeFunction.frame.endType,
+         fieldArgToChannel(windowNodeFunction.frame.startValue),
+         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+  }
+}
+
+void Window::addInput(RowVectorPtr input) {
+  // TODO : Is resize() or resizeAll() needed here ?
```

Comment:
Would you clarify the question? What is resizeAll?

Reply:
I think this is some old dead comment. Removed it.